### ISSUE TITLE
False positive "Function declaration obscured" error when a function is defined within the branches of a conditional

### ISSUE THREAD

**Issue by robertmaxton42** [microsoft/pyright#6018]
Suppose I define the function

```
def foo(test: bool):
    if test:
        def bar():
            return 1
    else:
        def bar():
            return 0
    return bar
```

`foo` behaves as one would expect, returning the constant function `()->Literal[1]` if its argument is true and the constant function `()->Literal[0]` otherwise. But Pylance thinks that the second definition of `bar` obscures the first, rather than simply being defined differently in different branches.

**VS Code extension or command-line**
VS Code extension v2023.9.21 (Pre-release)

**Comment by erictraut** [comment]
This isn't a false positive. You're declaring the symbol `bar` twice in incompatible ways, which is an error.

If you explicitly declare the return type to be `int` in both cases, the two declarations will be compatible, and the error will no longer be generated.

```python
def foo(test: bool):
    if test:

        def bar() -> int:
            return 1

    else:

        def bar() -> int:
            return 0

    return bar
```
